File: insets-helper/src/main/java/net/xpece/androidx/optical/OpticalInsetsHelper.java
package net.xpece.androidx.optical;

import android.annotation.TargetApi;
import android.graphics.Insets;
import android.graphics.drawable.Drawable;
import android.view.View;

import androidx.annotation.RequiresApi;

/**
 * Helper class for making views aware of optical insets
 * on {@link android.graphics.drawable.InsetDrawable InsetDrawable} below API 21,
 * and {@link android.graphics.drawable.LayerDrawable LayerDrawable} (including children
 * such as {@link android.graphics.drawable.RippleDrawable RippleDrawable}) on all API levels.
 */
@RequiresApi(16)
public final class OpticalInsetsHelper {

    private final View view;

    private Insets opticalInsets = null;

    public OpticalInsetsHelper(final View view) {
        this.view = view;
    }

    /**
     * Return this from {@code View#getOpticalInsets()}.
     */
    @TargetApi(29)
    public Insets onGetOpticalInsets() {
        if (opticalInsets == null) {
            // Same as platform, we don't support changing insets once resolved.
            final Drawable background = view.getBackground();
            opticalInsets = background != null
                    ? OpticalInsets.getOpticalInsetsCompat(background)
                    : InsetsCompat.NONE;
            // If background.getOpticalInsets() returns null this will blow up.
        }
        return opticalInsets;
    }

    /**
     * Call this from {@code View#setOpticalInsets(Insets)}.
     */
    public void onSetOpticalInsets(final Insets insets) {
        if (opticalInsets != insets) {
            opticalInsets = insets;
            view.requestLayout();
        }
    }
}

File: insets-helper/src/main/java/net/xpece/androidx/optical/OpticalInsetsHelper.kt
package net.xpece.androidx.optical

import android.graphics.Insets
import android.graphics.drawable.InsetDrawable
import android.graphics.drawable.LayerDrawable
import android.graphics.drawable.RippleDrawable
import androidx.annotation.RequiresApi
import android.view.View

/**
 * Helper class for making views aware of optical insets on [InsetDrawable] below API 21,
 * and [LayerDrawable] (including children such as [RippleDrawable]) on all API levels.
 */
@RequiresApi(16)
class OpticalInsetsHelper(private val view: View) {

    private var opticalInsets: Insets? = null

    /**
     * Return this from `View.getOpticalInsets()`.
     */
    fun onGetOpticalInsets(): Insets {
        if (opticalInsets == null) {
            // Same as platform, we don't support changing insets once resolved.
            opticalInsets = view.background?.getOpticalInsets() ?: InsetsCompat.NONE
        }
        return opticalInsets!!
    }

    /**
     * Call this from `View.setOpticalInsets(Insets)`.
     */
    fun onSetOpticalInsets(insets: Insets) {
        if (opticalInsets != insets) {
            opticalInsets = insets
            view.requestLayout()
        }
    }
}
